provider/docker/docker.go

```diff
@@ -155,47 +260,103 @@ func (p *Provider) Provide(configurationChan chan<- types.ConfigMessage, pool *s
 				Configuration: configuration,
 			}
 			if p.Watch {
-				ctx, cancel := context.WithCancel(ctx)
 				if p.SwarmMode {
 					errChan := make(chan error)
-					// TODO: This need to be change. Linked to Swarm events docker/docker#23827
-					ticker := time.NewTicker(SwarmDefaultWatchTime)
 					pool.Go(func(stop chan bool) {
+						watchCtx, cancel := context.WithCancel(ctx)
+						defer cancel()
+
 						defer close(errChan)
-						for {
-							select {
-							case <-ticker.C:
-								services, err := listServices(ctx, dockerClient)
+
+						callback := &EventCallback{
+							ListAndUpdateServicesFunc: func() error {
+								err := p.listAndUpdateServices(watchCtx, dockerClient, configurationChan)
```

Comment:
could be `return p.listAndUpdateServices(…)` — the log shouldn't be there (but in the code using this one). But really, I feel `EventCallback` shouldn't be there (or that way)